tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc

```diff
@@ -0,0 +1,230 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements nccl apis for alpa to use. 
+
+#include "tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.h"
+
+#include <memory>
+#include <utility>
+#include <stdexcept>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/str_format.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/time/time.h"
+#include "tensorflow/compiler/xla/debug_options_flags.h"
+#include "tensorflow/compiler/xla/service/global_device_id.h"
+#include "tensorflow/compiler/xla/service/gpu/gpu_executable_run_options.h"
+#include "tensorflow/compiler/xla/status_macros.h"
+#include "tensorflow/compiler/xla/statusor.h"
+#include "tensorflow/core/platform/env.h"
+
+#include "pybind11/pybind11.h"
+#include "pybind11/attr.h"
+#include "pybind11/cast.h"
+#include "pybind11/numpy.h"
+#include "pybind11/pytypes.h"
+#include "pybind11/stl_bind.h"
+
+#include "tensorflow/compiler/xla/python/py_buffer.h"
+#include "tensorflow/compiler/xla/python/py_client.h"
+
+PYBIND11_MAKE_OPAQUE(std::vector<ncclComm_t>);
+
+namespace xla {
+namespace alpa_nccl {
+
+ncclDataType_t MyToNcclDataType(PrimitiveType element_type) {
```

Comment:
Why do you need to implement a custom version of this function?

Reply:
Because ToNcclDatatype is inside anonymous namespace {} that could only be used in the file that it locates(nccl_utils.cc). I will change it to NcclDatatype.